### PR TITLE
Mastadon/Ecko Plugins

### PR DESCRIPTION
**Pull request by Mansh05**
This pr deals with adding new gem for plugin registry and configuring options on different levels. Here I have currently development a gem where people can simply use the registry and and the registry will register an engine which will help integrate mastodon instances in with different type of plugins. Currently in this PR I have developed the registry and the stripe plugin.

With this pr, I will perform few thing as follows.
1. Fix all build issues
2. Export gem to another repo. (Needs suggestions here)
3. And what all plugins to develop in future. (Teams and Groups or a moderation system)

This will now stream line the process on how we can build plugins for mastodon. The gems named for this instance is under ecko module.

In this POC, we have made a simple stripe based plugin which will help create checkout intents and pay accordingly. This plugin can be easily hooked on any type of workflow within mastodon, even on other rails apps instance.

Specs and will be updated within the process of this pr.

### CODE REVIEW

**Comment by weex** [comment]
Two things. One is removing the license file so that it falls under AGPL with the rest of the project. The other is that code of conduct file which I really like and can address https://github.com/magicstone-dev/magicstone.dev/issues/32. For that matter, it should be adopted by Magic Stone in that .dev repo or the wiki and linked from Ecko and Acropolis and C4-tools repos.

**Comment by Mansh05** [comment]
@weex Done with your suggestions.

File: app/models/stripe_payment_intent.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#  id           :bigint(8)        not null, primary key
+#  code         :string           not null
+#  reference    :string           not null
+#  category     :string           default("stripe"), not null
+#  payable_type :string
+#  payable_id   :bigint(8)
+#  metadata     :jsonb            not null
+#
+
+class StripePaymentIntent < ApplicationRecord
```
Comment (weex) [inline app/models/stripe_payment_intent.rb:15]: My understanding of a plugin would be that code doesn't need to be mixed into the main codebase. Would this be an exception to that or should this file be placed inside the gem?

File: gems/ecko-plugins/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).
+
+## Contributing
+
+Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/ecko-plugins. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.
+
+## License
+
+The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).
```
Comment (weex) [inline gems/ecko-plugins/README.md:39]: Not sure we need to suggest a license but if we do it should be GPL/AGPL or compatible.

File: gems/ecko-plugins/ecko-plugins.gemspec [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+Gem::Specification.new do |spec|
+  spec.name          = 'ecko-plugins'
+  spec.version       = Ecko::Plugins::VERSION
+  spec.authors       = ['Manish Sharma']
+  spec.email         = ['arnoltherasing@gmail.com']
+
+  spec.summary       = 'Adds plugins functionality to mastadon as a whole'
+  spec.description   = 'We can register to plugins through the registry provided by ecko plugins'
+  spec.homepage      = 'https://github.com/magicstone-dev/ecko'
+  spec.license       = 'MIT'
```
Comment (weex) [inline gems/ecko-plugins/ecko-plugins.gemspec:15]: Same about license

**Comment by weex** [comment]
Feel free to rename not WIP when ready to merge. Is it correct to say you're adding one gem for the plugin registry, and then the first plugin is staying in this codebase but simply under the /gems directory tree?

File: gems/ecko-plugins/ecko-plugins.gemspec (hunk 3 above)
Comment (Mansh05) [reply gems/ecko-plugins/ecko-plugins.gemspec:15]: Ah ok will change this. Thanks.

File: app/models/stripe_payment_intent.rb (hunk 1 above)
Comment (Mansh05) [reply app/models/stripe_payment_intent.rb:15]: Ah this is a file generated by the gem actually. This file is there in the generator of the gem.
we can run rails generator code to generate this code and migration with it.
rails generate ecko:plugins:stripe:payment_intention will generate this file.

**Comment by Mansh05** [comment]
Actually we can easily extract both the registry and the plugin as gem. Ecko plugins can be a sdk for plugins within the whole mastodon ecosystem and stripe can be one of the registered plugin for it. We can pull it out as two different gems when we have multiple plugins in future, I will change the title for the pr as well.

**Pull request merged**